manual/src/cmds/lexyacc.etex

```diff
@@ -724,4 +724,12 @@ positions inside the scanned lexer buffer. The size of this table is
 limited to at most 255 cells. This error should not show up in normal
 situations.
 
+\item[ocamlyacc: concurrency safety]
+
+Parsers generated by ocamlyacc are not generally concurrency safe.
+Those parsers rely on an internal work state which is specific to each
+domain but is shared between various threads.
+It is thus recommended to use locks when using ocamlyacc parsers
+on different threads, but domain parallelism is achievable.
+
```

Comment:
Could we state a little more explicitly what is safe?  If I understand this right, spawning multiple parsers in separate domains and letting them run to completion is safe.  Suspending a computation on a domain and resuming it on the same domain is also safe, if no other parsers run in the interim on that domain.

But stack switching and migrating the parser execution to a different domain is _not_ safe.  And suspending a parser execution on a domain and resuming another parser execution on the same domain is not safe.

Reply:
I agree that it is better to be more explicit, and yes the simple safe use case is to run a parser to completion on the same domain without any interleaving without another ocamlyacc parser.

Maybe something like:

> It is safe to spawn multiple parsers in separate domains and let them run to completion.
> It is also safe to suspend and resume a parser on the same domain if no other parsers run in the interim on this domain.
> It is not safe to migrate a parser execution to a different domain or to interleave the execution of two or more parsers on the same domain.

?

Reply:
That reads better to me!  It might also be worth giving a small example about how locks might be used to achieve concurrency-safety:

>  users will still need to use locks to achieve concurrency safety

It's not quite clear to me how to use locks safely here without a single lock around the entire execution of the parser.

Reply:
If I understand correctly, unsafe usage of ocamlyacc according to these guidelines is already possible in OCaml 4 (you can call recursively another parser from within a semantic action) and will not break memory- or type-safety, right?

I think the documentation could clarify what safe means here, and in particular what unsafe means. If I'm hunting a segfault in a Multicore OCaml program, should I or should I not rule out ocamlyacc as a plausible culprit?

Reply:
Unsafe usage is already possible with either inner call to other parsers or threads. The thread case can break type safety since it is possible to make one parser read the stack of another one with a mismatched type.

Reply:
This is a significant turn-off: it means that ocamlyacc was already memory-unsafe to use with OCaml 4.x, and this PR makes the code more complex (and potentially slightly less efficient) without fixing the issue.

Reply:
(Florian told me "by voice" about an alternative approach he has in mind that is simpler and may also fix this Threads issue.)

Reply:
Since after more thinking, we really don't want to encourage users to run ocamlyacc parser concurrently, a simpler fix is probably to add a global lock to `Parsing` and make `yyparse` acquire this lock. This creates a deadlock for nested ocamlyacc parsers (parser calling other parser in their semantic action), but this use case is not really supported and we could potentially add a global parameter to allow such nested parsers.

Reply:
If using nested parsers is not supported, a lock will detect it and fail appropriately, I think ("deadlock avoided"). The user should even get a nice stacktrace showing where the lock attempt failed?